src/Compiler.ts

```diff
@@ -11,42 +11,96 @@ import { LuaLibImportKind, LuaTarget, LuaTranspiler, TranspileError } from "./Tr
 
 export function compile(argv: string[]): void {
     const commandLine = parseCommandLine(argv);
-    compileFilesWithOptions(commandLine.fileNames, commandLine.options);
+    /* istanbul ignore if: tested in test/compiler/watchmode.spec with subproccess */
+    if (commandLine.options.watch) {
+        watchWithOptions(commandLine.fileNames, commandLine.options);
+    } else {
+        compileFilesWithOptions(commandLine.fileNames, commandLine.options);
+    }
 }
 
-export function compileFilesWithOptions(fileNames: string[], options: CompilerOptions): void {
-    if (!options.luaTarget) {
-        options.luaTarget = LuaTarget.LuaJIT;
+/* istanbul ignore next: tested in test/compiler/watchmode.spec with subproccess */
+export function watchWithOptions(fileNames: string[], options: CompilerOptions): void {
+    let host: ts.WatchCompilerHost<ts.SemanticDiagnosticsBuilderProgram>;
+    let config = false;
+    if (options.project) {
+        config = true;
+        host = ts.createWatchCompilerHost(
+            options.project,
+            options,
+            ts.sys,
+            ts.createSemanticDiagnosticsBuilderProgram
+        );
+    } else {
+        host = ts.createWatchCompilerHost(
+            fileNames,
+            options,
+            ts.sys,
+            ts.createSemanticDiagnosticsBuilderProgram
+        );
+    }
+
+    const errorDiagnostic: ts.Diagnostic = {
```

Comment:
I would prefer you create this and immediately fill in the correct values in `afterProgramCreate` since it's not used anywhere else.